Clarify time value: timezones + values spanning midnight (#328)
google/transit: gtfs/spec/en/reference.md

- **Float** - A floating point number.
- **Integer** - An integer.
- **Phone number** - A phone number.

> - **Time** - Time must be either in the HH:MM:SS or H:MM:SS format. The time must be measured in the hours from midnight on the day on which the trip began using the time zone specified by `agency.agency_timezone`. For trips which span midnight in the timezone specified by `agency.agency_timezone`, times after midnight should be greater than 24:00:00. <br> *Example: `14:30:00` for 2:30PM or `25:35:00` for 1:35AM on the next day.*

Reviewer: > on the day on which the trip began

With a trip having 3 `stop_times` at `25:00:00`, `25:30:00` & `26:00:00`, running on service day `2022-06-06`, would the time be measured from `2022-06-06T00:00:00+$offset` or from `2022-06-07T00:00:00+$offset`? I think the previous phrasing "service day" is more precise.

Reviewer: I also think removing the "noon minus 12h" definition a) breaks existing datasets, and b) doesn't make processing GTFS datasets easier and less error-prone. I proposed alternative changes here: https://github.com/google/transit/issues/325#issuecomment-1126346838

> - **Time** - Time must be either in the HH:MM:SS or H:MM:SS format. The time must be measured in the hours from midnight on the day on which the trip began using the time zone specified by `agency.agency_timezone`. For trips which span midnight in the timezone specified by `agency.agency_timezone`, times after midnight should be greater than 24:00:00. <br> *Example: `14:30:00` for 2:30PM or `25:35:00` for 1:35AM on the next day.*

Reviewer 2: ```suggestion
- **Time** - Time in the HH:MM:SS format (H:MM:SS is also accepted). The time is measured from "noon minus 12h" of the service day.

  (effectively midnight except for days on which daylight savings time changes occur). For times occurring after midnight, enter the time as a value greater than 24:00:00 in HH:MM:SS local time for the day on which the trip schedule begins. <br> *Example: `14:30:00` for 2:30PM or `25:35:00` for 1:35AM on the next day.*
```

|  `parent_station` | Foreign ID referencing `stops.stop_id` | **Conditionally Required** | Defines hierarchy between the different locations defined in `stops.txt`. It contains the ID of the parent location, as followed:<br><br>- **Stop/platform** (`location_type=0`): the `parent_station` field contains the ID of a station.<br>- **Station** (`location_type=1`): this field must be empty.<br>- **Entrance/exit** (`location_type=2`) or **generic node** (`location_type=3`): the `parent_station` field contains the ID of a station (`location_type=1`)<br>- **Boarding Area** (`location_type=4`): the `parent_station` field contains ID of a platform.<br><br>Conditionally Required:<br>- **Required** for locations which are entrances (`location_type=2`), generic nodes (`location_type=3`) or boarding areas (`location_type=4`).<br>- Optional for stops/platforms (`location_type=0`).<br>- Forbidden for stations (`location_type=1`).|

> |  `stop_timezone` | Timezone | Optional | Timezone of the location. If the location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with empty `stop_timezone` inherit the timezone specified by `agency.agency_timezone`. The times in [stop_times.txt](#stop_timetxt) must be entered as the time since midnight in the timezone specified by `agency.agency_timezone`. This ensures that the time values in a trip always increase over the course of a trip, regardless of which timezones the trip crosses. |

Reviewer: Even though this part of the definition has been in place before: It contradicts the GTFE *Time* definition.
Following this definition here, during DST end, they *would not* increase, right? e.g. `1:01:00` DST, `1:31:00` DST, `2:01:00` standard time.

> |  `stop_timezone` | Timezone | Optional | Timezone of the location. If the location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with empty `stop_timezone` inherit the timezone specified by `agency.agency_timezone`. The times in [stop_times.txt](#stop_timetxt) must be entered as the time since midnight in the timezone specified by `agency.agency_timezone`. This ensures that the time values in a trip always increase over the course of a trip, regardless of which timezones the trip crosses. |

Reviewer 2: In order to call attention to the requirement to enter times in the agency timezone, without contradicting the GTFS time definition, I suggest the following wording: 

```suggestion
|  `stop_timezone` | Timezone | Optional | Timezone of the location. If the location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with empty `stop_timezone` inherit the timezone specified by `agency.agency_timezone`. Note: Times in [stop_times.txt](#stop_timetxt) must always be entered in the timezone specified by`agency.agency_timezone`. This ensures that the time values in a trip always increase over the course of a trip, regardless of which timezones the trip crosses. |
```

> |  `stop_timezone` | Timezone | Optional | Timezone of the location. If the location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with empty `stop_timezone` inherit the timezone specified by `agency.agency_timezone`. The times in [stop_times.txt](#stop_timetxt) must be entered as the time since midnight in the timezone specified by `agency.agency_timezone`. This ensures that the time values in a trip always increase over the course of a trip, regardless of which timezones the trip crosses. |

Reviewer 2: In order to call attention to the requirement to enter times in the agency timezone, without contradicting the GTFS time definition, I suggest the following wording: 

```suggestion
|  `stop_timezone` | Timezone | Optional | Timezone of the location. If the location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with empty `stop_timezone` inherit the timezone specified by `agency.agency_timezone`. Note: Times in [stop_times.txt](#stop_timetxt) must always be entered in the timezone specified by`agency.agency_timezone`. This ensures that the time values in a trip always increase over the course of a trip, regardless of which timezones the trip crosses. |
```

|  Field Name | Type | Presence | Description |
|  ------ | ------ | ------ | ------ |
|  `trip_id` | Foreign ID referencing `trips.trip_id` | **Required** | Identifies a trip.  |
|  `arrival_time` | Time | **Conditionally Required** | Arrival time at the stop (defined by `stop_times.stop_id`) for a specific trip (defined by `stop_times.trip_id`). Time must be either in the HH:MM:SS or H:MM:SS format. The time must be measured in the hours from midnight on the day on which the trip began using the time zone specified by `agency.agency_timezone`. For trips which span midnight in the timezone specified by `agency.agency_timezone`, times after midnight should be greater than 24:00:00. <br><br>If there are not separate times for arrival and departure at a stop, `arrival_time` and `departure_time` should be the same. <br><br> If exact arrival and departure times (`timepoint=1` or empty) are not available, estimated or interpolated arrival and departure times (`timepoint=0`) should be provided.<br><br>Conditionally Required:<br>- **Required** for the first and last stop in a trip (defined by `stop_times.stop_sequence`). <br>- **Required** for `timepoint=1`.<br>- Optional otherwise.|

> |  `departure_time` | Time | **Conditionally Required** | Departure time from the stop (defined by `stop_times.stop_id`) for a specific trip (defined by `stop_times.trip_id`). Time must be either in the HH:MM:SS or H:MM:SS format. The time must be measured in the hours from midnight on the day on which the trip began using the time zone specified by `agency.agency_timezone`. For trips which span midnight in the timezone specified by `agency.agency_timezone`, times after midnight should be greater than 24:00:00.<br><br> If exact arrival and departure times (`timepoint=1` or empty) are not available, estimated or interpolated arrival and departure times (`timepoint=0`) should be provided.<br><br>Conditionally Required:<br>- **Required** for `timepoint=1`.<br>- Optional otherwise.| |

Reviewer 2: These definitions should be updated to match the definition of GTFS time. Perhaps instead of repeating the text, they could state that GTFS Time should be used?

> |  `departure_time` | Time | **Conditionally Required** | Departure time from the stop (defined by `stop_times.stop_id`) for a specific trip (defined by `stop_times.trip_id`). Time must be either in the HH:MM:SS or H:MM:SS format. The time must be measured in the hours from midnight on the day on which the trip began using the time zone specified by `agency.agency_timezone`. For trips which span midnight in the timezone specified by `agency.agency_timezone`, times after midnight should be greater than 24:00:00.<br><br> If exact arrival and departure times (`timepoint=1` or empty) are not available, estimated or interpolated arrival and departure times (`timepoint=0`) should be provided.<br><br>Conditionally Required:<br>- **Required** for `timepoint=1`.<br>- Optional otherwise.| |

Reviewer 2: These definitions should be updated to match the definition of GTFS time. Perhaps instead of repeating the text, they could state that GTFS Time should be used?
